_remake/lib/init-service-routes.js

```diff
@@ -0,0 +1,198 @@
+        return res.status(200).end();
+      });
+  });
+
+  // endpoint for registering subdomain (save in DB)
+  // validation callbacks: checkJWT, validSubdomain
+  // user must be authenticated to access it
+  app.post('/service/subdomain/register', checkJWT, validSubdomain, (req, res) => {
+    const { subdomain } = req.body;
+
+    connection.query('SELECT * FROM apps WHERE user_id = ?',
+      [req.user_id],
+      (err, results, fields) => {
+        if (err)
+          return res.status(500).json(err).end();
+        if (results.length >= global.config.limits.appPerUser)
+          return res.status(403).json({ message: `Reached ${global.config.limits.appPerUser} apps limit.` }).end();
+
+        connection.query('INSERT INTO apps (name, user_id, domain) VALUES (?, ?, ?)',
+          [subdomain, req.user_id, `${subdomain}.remakeapps.com`],
+          (err, results, fields) => {
+            if (err) {
+              if (err.code === "ER_DUP_ENTRY")
+                return res.status(400).json({ message: "An app with the same name already exists."}).end();
+              else return res.status(500).json(err).end();
+            }
+            return res.status(200).end();
+          });
+      });
+  });
+
+  // endpoint for deploying app
+  // upload app files if the user owns the app
+  // validation callbacks: checkJWT, validSubdomain, upload.single(...)
+  // user must be authenticated to access it
+  app.post('/service/deploy', checkJWT, upload.single('deployment'), validSubdomain, (req, res) => {
+    const { appName } = req.body;
+
+    connection.query('SELECT * FROM apps WHERE user_id = ? AND name = ?',
+      [req.user_id, appName],
+      (err, result, _) => {
+        if (err) return res.status(500).json(err).end();
+        if (result.length === 0) {
+          return res.status(403).json('Unauthorized to deploy').end();
+        }
+        extract(req.file.path, { dir: `${global.config.location.tmp}/${appName}` }, (err) => {
+          if (err) {
+            return res.status(500).json(err).end();
+          } else {
+            try {
+              shell.mkdir('-p', `${global.config.location.remake}/app/${appName}`);
+              shell.mkdir('-p', `${global.config.location.remake}/_remake-data/${appName}`);
+              shell.mkdir('-p', `${global.config.location.remake}/_remake-data/${appName}/user-app-data`);
+              shell.mkdir('-p', `${global.config.location.remake}/_remake-data/${appName}/user-details`);
+              shell.cp('-R', `${global.config.location.tmp}/${appName}/app/*`, `${global.config.location.remake}/app/${appName}/`);
+              // commented this out. do we want to deploy local development data?
+              // shell.cp('-R', `${global.config.location.tmp}/${appName}/_remake-data/*`, `${global.config.location.remake}/_remake-data/${appName}/`);
+              shell.rm(req.file.path);
+              shell.rm('-R', (req.file.path.split('.'))[0]);
```

Comment:
Are we **super, super, super** sure that this incoming file path won't and cannot be anything malicious? We're recursively removing directories here, so we might want to super confirm that it's the directory we absolutely want to remove.

Reply:
That's right. This may be a flaw and I will find a better approach

Reply:
Awesome